### PR TITLE
Classification EN Secondary (BERT / Bedrock) - Multiclass / Rank

### LINKED ISSUE #447: Classification of unconsolidated rock components (EN norm)
Extend the classifier for the unconsolidated rock lithology (main component) to include also any number (up to five) of secondary components.

There are 24 values for the main component currently available in the Boreholes database schema: [lithology_uncon_main_codelist.csv](https://github.com/user-attachments/files/26601495/lithology_uncon_main_codelist.csv)

There are 20 values for the secondary components (2 until 6) currently available in the Boreholes database schema: [lithology_uncon_secondary_codelist.csv](https://github.com/user-attachments/files/26601492/lithology_uncon_secondary_codelist.csv)

This classification is Prio 1 for unconsolidated rock.

There is a good amount of data available in the Boreholes database. Most of the data comes from the "swisstopo" workgroup (primarily from GeoQuat boreholes) and from the canton of Thurgau.

The labelled data needs to be taken with a grain of salt. While it is unlikely to be completely incorrect, it might not always be the most precise labelling, as many of these labels were automatically mapped from other classification schemas that might not have allowed for the same level of precision.

For the different components, we have the following statistics in the database:
- Main component - [lithology_uncon_main_stats.csv](https://github.com/user-attachments/files/26600806/lithology_uncon_main_stats.csv)
  - 95289 labelled layers from 12163 distinct boreholes
  - Most common labels in workgroup "swisstopo": `"gravel"` (22785x), `"silt"` (12625x), `"sand"` (9316x)
  - Most common labels in workgroup "Thurgau": `"silt"` (7063x), `"gravel"` (6137x), `"sand"` (4617x)
- 2nd component - [lithology_uncon_2_stats.csv](https://github.com/user-attachments/files/26600810/lithology_uncon_2_stats.csv)
  - 81849 labelled layers from 11505 distinct boreholes
  - Most common labels in workgroup "swisstopo": `"silty"` (30398x), `"clayey"` (11204x), `"sandy"` (10069x)
  - Most common labels in workgroup "Thurgau": `"silty"` (7599x), `"clayey"` (7334x), `"sandy"` (3413x)
- 3rd component - [lithology_uncon_3_stats.csv](https://github.com/user-attachments/files/26600814/lithology_uncon_3_stats.csv)
  - 57264 labelled layers from 10331 distinct boreholes
  - Most common labels in workgroup "swisstopo": `"sandy"` (15514x), `"gravelly"` (7471x), `"stony / with stones"` (6317x)
  - Most common labels in workgroup "Thurgau": `"sandy"` (4432x), `"gravelly"` (3447x), `"silty"` (2346x)
- 4th component - [lithology_uncon_4_stats.csv](https://github.com/user-attachments/files/26600819/lithology_uncon_4_stats.csv)
  - 25878 labelled layers from 7641 distinct boreholes
  - Most common labels in workgroup "swisstopo": `"stony / with stones"` (9251x), `"gravelly"` (3104x), `"sandy"` (2597x)
  - Most common labels in workgroup "Thurgau": `"gravelly"` (2230x), `"stony / with stones"` (1383x), `"medium gravelly"` (1127x)
- 5th component - [lithology_uncon_5_stats.csv](https://github.com/user-attachments/files/26600824/lithology_uncon_5_stats.csv)
  - 6414 labelled layers from 3026 distinct boreholes
  - Most common labels in workgroup "swisstopo": `"stony / with stones"` (2036x), `"blocky / with blocks"` (1359x), `"gravelly"` (104x)
  - Most common labels in workgroup "Thurgau": `"sandy"` (676x), `"stony / with stones"` (520x), `"coarse gravelly"` (514x)
- 6th component - [lithology_uncon_6_stats.csv](https://github.com/user-attachments/files/26600826/lithology_uncon_6_stats.csv)
  - 1327 labelled layers from 860 distinct boreholes
  - Most common labels in workgroup "swisstopo": `"blocky / with blocks"` (543x), `"stony / with stones"` (19x), `"with organic inclusion"` (18x)
  - Most common labels in workgroup "Thurgau": `"sandy"` (402x), `"coarse gravelly"` (127x), `"blocky / with blocks"` (62x)

A classifier for the main component was already trained in 2025 in https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/339 / https://github.com/swisstopo/swissgeol-boreholes-dataextraction/pull/344

If possible, it would be preferable to train one classifier that outputs an ordered list of classified components (of variable length). Because with independent classifiers for every possible component, we cannot really control that the final list of classified components is complete and does not contain repetitions.

The possible values for main component are different from the possible values from the secondary component. Still, the classification of main and secondary component is not independent at all. For example, is the main component is already "sand", then "sandy" should not be included among the secondary components.

```sql
WITH data AS (
SELECT
  id_bho,
  name_wgp,
  lithological_description.depth_from,
  (SELECT text_cli_en from bdms.codelist WHERE codelist.id_cli = lithology_description.lithology_uncon_main_id) as property,
  (SELECT STRING_AGG(identifier_value, ',') FROM bdms.borehole_identifiers_codelist WHERE identifier_id = 100000006 AND borehole_id = id_bho) as geoquat_id
FROM bdms.borehole
JOIN bdms.stratigraphy ON stratigraphy.borehole_id = borehole.id_bho
LEFT JOIN bdms.lithological_description ON
  stratigraphy.id = lithological_description.stratigraphy_id
LEFT JOIN bdms.lithology ON
  lithological_description.depth_from = lithology.depth_from AND 
  lithological_description.depth_to = lithology.depth_to AND
  lithology.stratigraphy_id = stratigraphy.id 
LEFT JOIN bdms.lithology_description ON
  lithology_description.lithology_id = lithology.id
JOIN bdms.workgroups ON id_wgp_fk = id_wgp
)

SELECT name_wgp, property, COUNT(*) as layers, COUNT(DISTINCT id_bho) as boreholes, COUNT(geoquat_id) as geoquat_layers FROM data
WHERE property IS NOT NULL
GROUP BY name_wgp, property
ORDER BY COUNT(*) DESC
```
(Replace `lithology_uncon_main_id` with `lithology_uncon_2_id` until `lithology_uncon_6_id` as desired.)

### PR DESCRIPTION
**Pull request by christianabbet**
Closes #447

What is left to do in PR:
[ ] Create a prompt for Bedrock to classifiy and rank components for current task. Baseline prompt only provide classification so far. Update tables below. Repot F1-micro/macro and Kendall score.
[ ] Add ranking mertirc in compute metric logic (use logic of `rank_metrics_from_layers()`. Maybe the function can be refactored to two functions with `rank_metrics_from_layers` that calls `rank_metrics_from(predictions, labels)` that can aslo be called from compute metric)

|           | Bedrock (v1) |          |               | BERT (Rank) |          |               |
|-----------|--------------|----------|---------------|-------------|----------|---------------|
| Test set  | F1-macro     | F1-micro | Kendall's Tau | F1-macro    | F1-micro | Kendall's Tau |
| Deepwells | 0.766           | 0.966            | 0.832                  | 0.629       | 0.966    | 0.965         |
| Geoquat   |  0.736            | 0.9196            | 0.621        | 0.829       | 0.948    | 0.778         |
| Nagra     | 0.668            | 0.963            | 0.866         | 0.880       | 0.991    | 0.798         |
| Thurgau   | 0.581            | 0.885            |  0.595           | 0.646       | 0.932    | 0.617         |
| Overall   | -            | -        | -             | 0.808       | 0.945    | 0.744         |

# BERT

| Loss | Test set  | BERT F1-macro | BERT F1-micro | BERT Kendall's Tau |
|------|-----------|---------------|---------------|--------------------|
| XEnt | Deepwells | 0.696         | 0.979         | -0.096             |
|      | Geoquat   | 0.840         | 0.951         |  0.244             |
|      | Nagra     | 1.000         | 1.000         |  0.195             |
|      | Thurgau   | 0.650         | 0.933         |  0.202             |
|      | Overall   | 0.854         | 0.947         |  -                 |
| Rank | Deepwells | 0.629         | 0.966         |  0.965             |
|      | Geoquat   | 0.829         | 0.948         |  0.778             |
|      | Nagra     | 0.880         | 0.991         |  0.798             |
|      | Thurgau   | 0.646         | 0.932         |  0.617             |
|      | Overall   | 0.808         | 0.945         |  0.744             |

# Bedrock

| Prompt   | Test set  | n files | Bedrock F1-macro | Bedrock F1-micro | Bedrock Kendall's Tau |
|----------|-----------|-----|------------------|------------------|-----------------------|
| Baseline | Deepwells | 3|  0.766            | 0.966            |  0.499                     |
|          | Geoquat   | 845| 0.649            | 0.933            | 0.249                   |
|          | Nagra     | 17 |0.571            | 0.954            | 0.899                   |
|          | Thurgau   | 644 | 0.575            | 0.895            | 0.14                   |
|          | Overall   | |-                | -                | -                     |
| v1       | Deepwells | 3 | 0.766           | 0.966            | 0.832                 |
|          | Geoquat   | 845 | 0.736            | 0.9196            | 0.621                 |
|          | Nagra     |  17 | 0.668            | 0.963            | 0.866                 |
|          | Thurgau   | 644 |0.581            | 0.885            |  0.595                 |
|          | Overall   | -                | -                | -                     |

# Examples `Thurgau`

| Doc ID | Description | Bedrock | BERT | GT | 
|--------|-------------|---------|------|----|
|  10778 | Limon sablo-argileux, beige, dur (à ferme), avec quelques cailloux et blocaux mal distribués.\nNombreux coquillages et esquilles noirâtres.\nNappe phréatique à 7.20 m (avril 1995). | ... | si > cl > co > sa > bo | si > cl > sa > co > bo |
|   1095 | Gravier sablo-limoneux beige, calcaire. | ... | gr > si > sa | gr > si > sa |
|  11255 | Limon argileux brun beige, avec quelques débris de brique, parfois finement sableux, tendre plastique. | ... | ba > si > cl > sa | ba > si > cl > fsa |
|   A102 | siltiger Kies mit viel Sand, nass, gelbgrau und grau | ... | gr > si > ca | gr > si > ca |
|  A1072 | Ausbildung wie oberer Horizont. Moräne, sandig-kiesiger Typ. Unterer Horizont | ... | sa > si > gr > cl | gr > si > sa |
| A11060 | toniger Silt mit wenig Sand und Kies, steif bis halbfest, dunkelgrau, nur mit Steinen, Korn eckig, halbfest bis fest, dunkelgrau | ... | si > cl > sa > gr > co | si > cl > sa > gr > co > bo |

### CODE REVIEW

**Comment by AgathaSchmidt** [comment]
The work you did looks good 
Please "review" my changes as well @christianabbet

File: docs/Experiment_Results.md [hunk 1]
```diff
@@ -74,13 +74,52 @@
...
 
+## EN Secondary
+
+|           | Bedrock (v1) |          |               | BERT (Rank) |          |               |
+|-----------|--------------|----------|---------------|-------------|----------|---------------|
+| Test set  | F1-macro     | F1-micro | Kendall's Tau | F1-macro    | F1-micro | Kendall's Tau |
+| Deepwells | 0.xxx        | 0.xxx    | 0.xxx         | 0.629       | 0.966    | 0.965         |
+| Geoquat   | 0.xxx        | 0.xxx    | 0.xxx         | 0.829       | 0.948    | 0.778         |
+| Nagra     | 0.xxx        | 0.xxx    | 0.xxx         | 0.880       | 0.991    | 0.798         |
+| Thurgau   | 0.xxx        | 0.xxx    | 0.xxx         | 0.646       | 0.932    | 0.617         |
```
Comment (christianabbet) [inline docs/Experiment_Results.md]: Can you update these values for Bedrock ?

File: docs/Experiment_Results.md [hunk 2]
```diff
@@ -74,13 +74,52 @@
...
+| Baseline | Deepwells | 0.768            | 0.970            | -                     |
+|          | Geoquat   | 0.649            | 0.933            | -                     |
+|          | Nagra     | 0.514            | 0.954            | -                     |
+|          | Thurgau   | 0.575            | 0.895            | -                     |
+|          | Overall   | -                | -                | -                     |
+| v1       | Deepwells | 0.xxx            | 0.xxx            | 0.xxx                 |
+|          | Geoquat   | 0.xxx            | 0.xxx            | 0.xxx                 |
+|          | Nagra     | 0.xxx            | 0.xxx            | 0.xxx                 |
+|          | Thurgau   | 0.xxx            | 0.xxx            | 0.xxx                 |
+|          | Overall   | -                | -                | -                     |
```
Comment (christianabbet) [inline docs/Experiment_Results.md:107]: Same here - to update

File: src/classification/config/bedrock/bedrock_config_en_secondary.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Config file for material classification into EN secondary classes
+# Prompt file and version
+prompts_file: bedrock/prompts/bedrock_en_secondary_prompts.yml
+prompt_version: baseline
```
Comment (christianabbet) [inline src/classification/config/bedrock/bedrock_config_en_secondary.yml]: Shouldn't it be v1 ?

File: src/classification/models/train.py [hunk 4]
```diff
@@ -457,12 +526,27 @@ def compute_metrics(self, eval_pred: EvalPrediction) -> dict[str, float]:
...
+                for logit in logits
+            ]
+            label_enum_lists = [
+                [self._id2class_enum[i] for i in np.argsort(rank_row) if rank_row[i] < n_classes]
+                for rank_row in rank_labels
+            ]
+            metrics["kendall_tau"] = rank_metrics_from(pred_enum_lists, label_enum_lists)
+
+        return metrics
+
```
Comment (christianabbet) [inline src/classification/models/train.py:549]: No way to use : `rank_metrics_from` ?

Comment (AgathaSchmidt) [reply src/classification/models/train.py:549]: I am using `rank_metrics_from` in line 546

Comment (christianabbet) [reply src/classification/models/train.py:549]: Sorry, tired :). All good then

**Comment by christianabbet** [comment]
LGTM !
